pkg/controller/plan/adapter/base/mac_conflicts.go

```diff
@@ -0,0 +1,109 @@
+package base
+
+import (
+	"fmt"
+
+	"github.com/kubev2v/forklift/pkg/apis/forklift/v1beta1/ref"
+	webbase "github.com/kubev2v/forklift/pkg/controller/provider/web/base"
+	"github.com/kubev2v/forklift/pkg/controller/provider/web/ocp"
+	liberr "github.com/kubev2v/forklift/pkg/lib/error"
+	cnv "kubevirt.io/api/core/v1"
+)
+
+// MacConflict represents a MAC address conflict between source and destination VMs
+type MacConflict struct {
+	// MAC address that is conflicting
+	MAC string
+	// Destination VM that has the conflicting MAC address
+	DestinationVM string
+}
+
+// DestinationVM represents a destination VM with MAC addresses for conflict checking
+type DestinationVM struct {
+	Namespace string
+	Name      string
+	MACs      []string
+}
+
+// ExtractMACsFromInterfaces is a helper that extracts MAC addresses from KubeVirt interface slices
+// This reduces the inner loop duplication across providers
+func ExtractMACsFromInterfaces(interfaces []cnv.Interface) []string {
+	var macs []string
+	for _, iface := range interfaces {
+		macs = append(macs, iface.MacAddress)
+	}
+	return macs
+}
+
+// FindSourceVM retrieves a source VM by reference using the common pattern
+// This encapsulates the vm := &Type{} + Find() + error wrapping pattern
+func FindSourceVM[T any](inventory interface {
+	Find(interface{}, ref.Ref) error
+}, vmRef ref.Ref) (*T, error) {
+	vm := new(T)
+	err := inventory.Find(vm, vmRef)
+	if err != nil {
+		return nil, liberr.Wrap(err, "vm", vmRef.String())
+	}
+	return vm, nil
+}
+
+// GetDestinationVMsFromInventory retrieves and extracts destination VMs using the common pattern
+// This works for all providers (vsphere, ovirt, openstack, ova, ocp)
+func GetDestinationVMsFromInventory(client interface {
+	List(interface{}, ...webbase.Param) error
```

Comment:
Don't like having generic `interface`, the `client` should be type `web.Client`

Reply:
If you are having problems with cyclic dependencies, please create new interface in top of the file and reference which interface you are using in commnet. Or try to move the interface to separate module.